File: examples/python/quickstart.py
#!/usr/bin/env python3
import os
import sys
import time
import requests
from datetime import datetime, timezone, timedelta

BASE_URL = "https://cloud.sycon.io"
USERNAME = os.getenv("SYCON_USERNAME", "your_username")
PASSWORD = os.getenv("SYCON_PASSWORD", "your_password")

# Time window (last 24h)
END = datetime.now(timezone.utc).replace(microsecond=0)
START = END - timedelta(days=1)

DEVICE_ID = 12345  # replace with your own
FIELD = "TEMPERATURE_CELSIUS"  # see enum in OpenAPI

def iso(dt: datetime) -> str:
    return dt.isoformat().replace("+00:00", "Z")

def login():
    r = requests.post(
        f"{BASE_URL}/auth/login",
        json={"username": USERNAME, "password": PASSWORD},
        timeout=20,
    )
    if r.status_code != 200:
        print("Login failed:", r.status_code, r.text)
        sys.exit(1)
    auth = r.headers.get("Authorization") or ""
    renew = r.headers.get("Renew") or ""
    if not auth or not renew:
        print("Missing Authorization/Renew headers:", dict(r.headers))
        sys.exit(1)
    jwt = auth.split(" ", 1)[1] if auth.startswith("Bearer ") else auth
    return jwt, renew

def auth_check(jwt: str):
    r = requests.get(
        f"{BASE_URL}/auth/check",
        headers={"Authorization": f"Bearer {jwt}"},
        timeout=10,
    )
    print("auth/check:", r.status_code)

def list_devices(jwt: str):
    r = requests.get(
        f"{BASE_URL}/api/devices",
        headers={"Authorization": f"Bearer {jwt}", "Accept": "application/json"},
        timeout=20,
    )
    r.raise_for_status()
    return r.json()

def get_raw(jwt: str, device_id: int, field: str, start: datetime, end: datetime, tail_limit=1000, external_sensor_id=None):
    params = {"start": iso(start), "end": iso(end), "tailLimit": tail_limit}
    if external_sensor_id:
        params["externalSensorId"] = external_sensor_id
    r = requests.get(
        f"{BASE_URL}/api/devices/{device_id}/{field}/data/raw",
        headers={"Authorization": f"Bearer {jwt}", "Accept": "application/json"},
        params=params,
        timeout=30,
    )
    print("raw data:", r.status_code)
    r.raise_for_status()
    return r.json()

def renew(jwt_renew: str):
    r = requests.get(f"{BASE_URL}/auth/renew", headers={"Renew": jwt_renew}, timeout=10)
    if r.status_code == 200:
        auth = r.headers.get("Authorization", "")
        new_jwt = auth.split(" ", 1)[1] if auth.startswith("Bearer ") else auth
        return new_jwt
    print("Renew failed:", r.status_code, r.text)
    return None

if __name__ == "__main__":
    jwt, renew_token = login()
    auth_check(jwt)

    devices = list_devices(jwt)
    print("Devices:", devices[:1], " ... total:", len(devices))

    data = get_raw(jwt, DEVICE_ID, FIELD, START, END, tail_limit=1000)
    print("First timestamps:", data.get("firstTimestamp"), "Last:", data.get("lastTimestamp"), "Count:", data.get("count"))

    # Optional: try renewing
    new_jwt = renew(renew_token)
    if new_jwt:
        print("JWT renewed (len:", len(new_jwt), ")")

File: examples/js/quickstart.js
// Run with: node examples/js/quickstart.js
// Requires Node 18+ (global fetch)

const BASE_URL = "https://cloud.sycon.io";
const USERNAME = process.env.SYCON_USERNAME || "your_username";
const PASSWORD = process.env.SYCON_PASSWORD || "your_password";

const DEVICE_ID = 12345; // replace with your own
const FIELD = "TEMPERATURE_CELSIUS"; // see enum in OpenAPI

function iso(dt) {
  return dt.toISOString().replace(/\.\d{3}Z$/, "Z");
}

async function login() {
  const r = await fetch(`${BASE_URL}/auth/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: USERNAME, password: PASSWORD }),
  });
  if (!r.ok) {
    const body = await r.text();
    throw new Error(`Login failed: ${r.status} ${body}`);
  }
  const auth = r.headers.get("Authorization") || "";
  const renew = r.headers.get("Renew") || "";
  if (!auth || !renew) {
    throw new Error("Missing Authorization/Renew headers");
  }
  const jwt = auth.startsWith("Bearer ") ? auth.slice(7) : auth;
  return { jwt, renew };
}

async function authCheck(jwt) {
  const r = await fetch(`${BASE_URL}/auth/check`, {
    headers: { Authorization: `Bearer ${jwt}` },
  });
  console.log("auth/check:", r.status);
}

async function listDevices(jwt) {
  const r = await fetch(`${BASE_URL}/api/devices`, {
    headers: { Authorization: `Bearer ${jwt}`, Accept: "application/json" },
  });
  if (!r.ok) throw new Error(`devices failed: ${r.status}`);
  return r.json();
}

async function getRaw(jwt, deviceId, field, start, end, tailLimit = 1000, externalSensorId) {
  const url = new URL(`${BASE_URL}/api/devices/${deviceId}/${field}/data/raw`);
  url.searchParams.set("start", start);
  url.searchParams.set("end", end);
  url.searchParams.set("tailLimit", String(tailLimit));
  if (externalSensorId) url.searchParams.set("externalSensorId", externalSensorId);

  const r = await fetch(url, {
    headers: { Authorization: `Bearer ${jwt}`, Accept: "application/json" },
  });
  console.log("raw data:", r.status);
  if (!r.ok) throw new Error(`raw data failed: ${r.status}`);
  return r.json();
}

async function renewJwt(renewToken) {
  const r = await fetch(`${BASE_URL}/auth/renew`, {
    headers: { Renew: renewToken },
  });
  if (r.ok) {
    const auth = r.headers.get("Authorization") || "";
    return auth.startsWith("Bearer ") ? auth.slice(7) : auth;
  }
  console.warn("Renew failed:", r.status);
  return null;
}

(async () => {
  const { jwt, renew } = await login();
  await authCheck(jwt);

  const devices = await listDevices(jwt);
  console.log("Devices (first):", devices[0], "total:", Array.isArray(devices) ? devices.length : 0);

  // Last 24h window
  const end = new Date();
  const start = new Date(end.getTime() - 24 * 60 * 60 * 1000);
  const data = await getRaw(jwt, DEVICE_ID, FIELD, iso(start), iso(end), 1000);
  console.log("Data count:", data.count, "First:", data.firstTimestamp, "Last:", data.lastTimestamp);

  const newJwt = await renewJwt(renew);
  if (newJwt) console.log("JWT renewed (len):", newJwt.length);
})().catch(err => {
  console.error(err);
  process.exit(1);
});
